### DISCUSSION TITLE
Q: Adding a Style to the Conditional Styles for Node With a Hotkey?

### DISCUSSION THREAD

**Discussion by autoca** [freeplane/freeplane#700] [Ideas]
Hi all,

I would like to ask if I did overlook anything, either an already existing feature in FP or an already existing script here in the forums, regarding the following.

I would like to be able to assign a style like, e.g., the "Important" style to a node "on top of" any style already assigned to that node.
In the end the node would then have two or more styles assigned to it at the same time (and I am not even talking about the automatic level styles).

Now, I _do_ know that this _is_ already possible with the conditional styles. However, as far as I know, this must be done by "managing the conditional styles" for that node in its respective dialog. This means that there is a considerable amount of work involved each time that I want to apply a style in such a way.

I am wondering if there already exists a solution that let's me apply, e.g., the "Important" style quickly with a hotkey on top of any other style(s) already applied. - And of course, the same hotkey should also remove that style when it is already applied.

If not, then technically, maybe, might it be possible with a script to add a conditional style with the condition "Always" as the top line in the conditional styles for node definition? And if it is already there, to remove it? Or does that go beyond the scripting capabilities?

MTIA and regards, a.

**Comment by macmarrum** [comment]
Some time ago I created a script to "move" currently assigned style to the Node Conditional Style list: 
- https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/moveStyleToNodeCondies.groovy

and another to move it back:
- https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/moveLastStyleFromNodeCondies.groovy

At the time I had little understanding how to list styles to choose from, therefore I opted for a two-step process: first apply a style to a node (e.g. using Command Search) then execute [moveStyleToNodeCondies.groovy](https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/moveStyleToNodeCondies.groovy)

Now that I've learned more about map styles, I'm planning to create a script to present a list of styles and add the chosen one to the Node Conditional Styles list.

**Reply by autoca** [reply]
Wow, nice plans, thank you!
Looking foreward to it happily.
I will try to work with your current scripts in the meantime.
Thank you very much!

**Reply by autoca** [reply]
I tried to tamper a bit with the code and want to insert the conditional style as the first one.
I found the line with addConditionalStyle but that obviously adds it at the end of the cond style table.
How do I add it at the beginning or move it to the beginning? Thx.

**Reply by dpolivaev** [reply]
I looked at the scripts. I think that they are likely to break sooner or later because they use internal API that can change at any time.
Therefore the next step after creating such scripts could be submitting poll requests to extend Freeplane API so that you get a "legal" way to access the methods.

Currently, the internal API has only the addConditionalStyle method which adds the style at the end.
Obviously, it could be easily extended to insert the new style at any position. Feel free to submit a PR for it too.

**Reply by autoca** [reply]
Thanks for looking into this, Dimitry,

By PR you mean a "pull request"?
Would I not have to provide code that you could pull?
I would rather submit another FR (feature request). ;-)

Is there a current "legal" API way how to move cond styles in the table up or down?

**Reply by dpolivaev** [reply]
Probably not.
You probably need to cooperate with @macmarrum.
My own resources are limited.

**Reply by macmarrum** [reply]
PR created for the internal API
- https://github.com/freeplane/freeplane/pull/705

**Reply by macmarrum** [reply]
@autoca Once PR#705 gets merged and a preview released, you can use the script
- https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/moveStyleToNodeCondiesBeginning.groovy

There's more work needed to extend the scripting API. The internal methods to add/insert a node conditional style allow to specify a condition. So far I haven't been able figured out how to design it so that the condition can be represented in the scripting API.

**Reply by autoca** [reply]
WOW, thank you, macmarrum, for this effort. Wow.

**Reply by dpolivaev** [reply]
I merged the PR, although it changes only the internal API, and therefore the scripts which use it can stop working at any time if it is changed again.

**Reply by macmarrum** [reply]
I am working on extending the scripting API

**Comment by autoca** [comment]
So far all this works awesome, thank you all.
Now, I want to insert a specific, known style by name as the first style in the node conditional styles table.
@macmarrum, I used your script "moveStyleToNodeCondiesBeginning.groovy" as a template: https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/moveStyleToNodeCondiesBeginning.groovy

However, I am stuck at the line
`(LogicalStyleController.controller as MLogicalStyleController).insertConditionalStyle(map, condiStyleModel, 0, true, null, iStyle, false)`

Obviously, I must somehow get an existing style by name in to the iStyle variable. But how do I do that?
I looked through here: file:///C:/Program%20Files/Freeplane/doc/api/index.html but did not recognize anything useful for me.

MTIA and regards, a.

**Reply by macmarrum** [reply]
If you have a node with the style applied, you can get it by `n.style.style`, where `n` is your node. This method uses the scripting API.

E.g. given the node with `ID_123456` has your "Custom style 1" applied
```groovy
def style = ID_123456.style.style
```

Otherwise, you might need to use internal API: 
```groovy
import org.freeplane.features.styles.StyleFactory
StyleFactory.create('Custom style 1')
```
If your style name is translated (a predefined style name expressed in a language other than English), you need to additionally use TranslatedObject for your style name.

**Reply by autoca** [reply]
Thank you, yes the part with n.style.style seems to be recognizable in your script.

Unfortunately the second part - which seems to be what I want to do - does not work for me.
So far, I have this code, trying to put the style 'Important' directly into the node conditional styles table as first item:

```
import org.freeplane.api.Node as FPN
import org.freeplane.features.map.MapModel
import org.freeplane.features.map.NodeModel
import org.freeplane.features.styles.ConditionalStyleModel
import org.freeplane.features.styles.IStyle
import org.freeplane.features.styles.LogicalStyleController
import org.freeplane.features.styles.mindmapmode.MLogicalStyleController
import org.freeplane.features.styles.StyleFactory

// from org.freeplane.features.styles.mindmapmode.ManageNodeConditionalStylesAction.getConditionalStyleModel
static ConditionalStyleModel getConditionalStyleModel(NodeModel node) {
    ConditionalStyleModel conditionalStyleModel = (ConditionalStyleModel) node.getExtension(ConditionalStyleModel.class)
    if (conditionalStyleModel == null) {
        conditionalStyleModel = new ConditionalStyleModel()
        node.addExtension(conditionalStyleModel)
    }
    return conditionalStyleModel
}

for (FPN node in c.selecteds) {
    IStyle iStyle = StyleFactory.create('Important')
    
    NodeModel nodeModel = node.delegate
    MapModel map = node.mindMap.delegate
    ConditionalStyleModel condiStyleModel = getConditionalStyleModel(nodeModel)
    (LogicalStyleController.controller as MLogicalStyleController).insertConditionalStyle(map, condiStyleModel, 0, true, null, iStyle, false)
}
menuUtils.executeMenuItems(['ResetStyleAction',])
```

**Reply by macmarrum** [reply]
I think that predefined styles use "key" and "translated name", and "Important" is the translated name.
What do you get as a result of `=node.style.name` on a node with "Important" applied?
I suppose "styles.important" – try supplying it to StyleFactory.

**Reply by autoca** [reply]
Hey, thx for your continous help,
I did a node.text = node.style.name on a node with the "Important" style.
And yes, the name is really 'styles.important'. That is true.
Unfortunately, it still does not work when I replace `IStyle iStyle = StyleFactory.create('Important')` with `IStyle iStyle = StyleFactory.create('styles.important')`.
The result is as in the picture:

![grafik](https://user-images.githubusercontent.com/37442404/193659406-9af0fa25-2ce5-4e0d-8e7f-ca417c93a4bb.png)

I actually had a different user style on the node itself (main style) and it changed in the tool panel to 'styles.important'. The node condi also changed to 'styles.important' in the tool panel but the manage conditional styles for node dialog shows 'Default'. The visuals of the actual node in the map changed to look as if it had just the normal level style applied to it but nothing else.

P.S.: What does the last line `menuUtils.executeMenuItems(['ResetStyleAction',])` actually do? I could not figure it out other than calling some existing menu command. But which one?

**Reply by macmarrum** [reply]
But of course it doesn't work. I forgot about TranslatedObject :unamused: 
```groovy
IStyle iStyle = StyleFactory.create(new TranslatedObject('styles.important'))
```
___
ResetStyleAction is `Format->Apply style->No style`. I use it to force an update after internal API is used to manipulate Conditional Styles.

If you use `Tools->Developer Tools->Menu item info`, you'll get

- Menu item: No style
- Key: **ResetStyleAction**
- Icon Key: ResetStyleAction.icon
- Command line option (copied to clipboard): -XResetStyleAction
- Path: //main_menu/format//styleActions//ResetStyleAction

**Reply by autoca** [reply]
Thx, but now it says:
![grafik](https://user-images.githubusercontent.com/37442404/193683468-defc28fa-7187-47f6-9eb0-cebba6870f32.png)

Do I maybe need another import?

**Reply by macmarrum** [reply]
You probably need to `import org.freeplane.core.resources.TranslatedObject` or use `new org.freeplane.core.resources.TranslatedObject(...)`

**Reply by autoca** [reply]
Ok, I now got it to work.
Still have to look over the code again before I post it, but as an ugly prototype it works.
Thank you for your help, macmarrum.

**Reply by autoca** [reply]
Ok, here is the code that seems to work with the "Important" style, respectively "styles.important".
```
// @ExecutionModes({ON_SINGLE_NODE="/menu_bar/MyTools/Syles"})
/*
  * Inserts the style 'styles.important' as the first Node Conditional Style
  */

import org.freeplane.api.Node as FPN
import org.freeplane.features.map.MapModel
import org.freeplane.features.map.NodeModel
import org.freeplane.features.styles.ConditionalStyleModel
import org.freeplane.features.styles.IStyle
import org.freeplane.features.styles.LogicalStyleController
import org.freeplane.features.styles.mindmapmode.MLogicalStyleController
import org.freeplane.features.styles.StyleFactory
import org.freeplane.core.resources.TranslatedObject

// from org.freeplane.features.styles.mindmapmode.ManageNodeConditionalStylesAction.getConditionalStyleModel
static ConditionalStyleModel getConditionalStyleModel(NodeModel node) {
    ConditionalStyleModel conditionalStyleModel = (ConditionalStyleModel) node.getExtension(ConditionalStyleModel.class)
    if (conditionalStyleModel == null) {
        conditionalStyleModel = new ConditionalStyleModel()
        node.addExtension(conditionalStyleModel)
    }
    return conditionalStyleModel
}

for (FPN node in c.selecteds) {
    IStyle curIStyle = node.style.style

    IStyle iStyle = StyleFactory.create(new TranslatedObject('styles.important'))

    NodeModel nodeModel = node.delegate
    MapModel map = node.mindMap.delegate
    ConditionalStyleModel condiStyleModel = getConditionalStyleModel(nodeModel)
    (LogicalStyleController.controller as MLogicalStyleController).insertConditionalStyle(map, condiStyleModel, 0, true, null, iStyle, false)

    if (curIStyle === null) {
        menuUtils.executeMenuItems(['ResetStyleAction',])
    } else {
        node.style.name = node.style.name
    }
}
```
Two remarks:
1) Unfortunately, the visuals do not get updated automatically when I undo (Ctrl+z) the whole thing.

~~2) I also have a real user style named "Question" with that I want to do the same thing. Unfortunately the code does not work with that "Question" style. I tried changing `IStyle iStyle = StyleFactory.create(new TranslatedObject('styles.important'))` to `IStyle iStyle = StyleFactory.create(new TranslatedObject('Question'))` and then also to `IStyle iStyle = StyleFactory.create(new 'Question')` but neither worked.
Any ideas? Would be much apprechiated.~~

MTIA and regards, a.

**Reply by autoca** [reply]
Ok, I overlooked a small word but the computer did not:
Of course it must be `IStyle iStyle = StyleFactory.create('Question')`. Now it works also for that real user style.

However, the first point is even worse. I found out that the visual appearance does also not change automatically on a redo (Ctrl+y). That is if I do an undo (Ctrl+z) it does not visually change back. I have to navigate to a different node for it to update visually. And if I then do a redo (Ctrl+y) it is the same thing reversed.

Just a small bada bumm, but still, please, help ... many, many thx.

**Comment by autoca** [comment]
@macmarrum, I now also adapted your other script https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/moveFirstStyleFromNodeCondies.groovy to just remove the first style from the node conditionals and do nothing else.

However, neither your script nor my adaption can be undone (Ctrl+z) at all. Not even with thereafter navigating to another node are the changes undone that the scripts cause.

Here is my script, but the adaptions are minute:
```
// @ExecutionModes({ON_SINGLE_NODE="/menu_bar/MyTools/Syles"})
/*
 * Removes the first style from Node Conditional Styles
 */

import org.freeplane.api.Node as FPN
import org.freeplane.features.map.NodeModel
import org.freeplane.features.styles.ConditionalStyleModel
import org.freeplane.features.styles.LogicalStyleController

c.selecteds.each {FPN selected ->
    NodeModel selectedModel = selected.delegate
    ConditionalStyleModel selectedConditionalStyleModel = selectedModel.getExtension(ConditionalStyleModel.class)
    if (selectedConditionalStyleModel && selectedConditionalStyleModel.styles.size() > 0) {
        def item = LogicalStyleController.controller.removeConditionalStyle(selectedConditionalStyleModel, 0)
        // selected.style.name = item.style.toString()
        selected.style.name = selected.style.name
    }
}

```

**Reply by macmarrum** [reply]
@autoca That's great. Congratulations on creating scripts that fit your needs.

Regarding `Ctrl+Z`, I have fixed an issue in my scripts (to use MLogicalStyleController). Please feel free to try them out. `Ctrl+Z` should work now.

**Reply by autoca** [reply]
Thanks for still helping.
Unfortunately, from my tests with moveStyleToNodeCondiesBeginning.groovy (https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/moveStyleToNodeCondiesBeginning.groovy) the new changes kind of broke Ctrl+z. It worked better before.

**Reply by macmarrum** [reply]
Thanks for trying it out and reporting back. This helped me to identify a bug in the recently added code, in the undo action. I'll create a fix soon.

**Reply by macmarrum** [reply]
I have create a PR, which will fix undo/redo -- I'll update my scripts soon
- #726

**Comment by macmarrum** [comment]
I have created a PR to add ConditionalStyles-related api
- https://github.com/freeplane/freeplane/pull/788

Once reviewed by Dimitry and merged (if no comments), the actions will be simplified and will be safeguarded against changes in the internal API.

**Reply by dpolivaev** [reply]
Thanks for the PR.
I will release Freeplane 1.10.5 soon and consider your changes in the following versions.
The next release should include new layout variants and get the number 1.11.1.
I will review your changes after I finish my currently planned implementation steps.

**Reply by macmarrum** [reply]
@dpolivaev this is just to bump the topic up in your mailbox

**Reply by dpolivaev** [reply]
Yes, I have too many balls in the air. I am sorry.
